File: 1730-shortest-path-to-get-food/1730-shortest-path-to-get-food.java
class Solution {
    public int getFood(char[][] grid) {
        Queue<int[]> q = new LinkedList<>();
        
        int res = 0;
        int r = grid.length;
        int c = grid[0].length;
        boolean[][] visited = new boolean[r][c];
        
        for(int i = 0; i < r; i++) {
            for(int j = 0; j < c; j++) {
                if (grid[i][j] == '*') {
                    q.offer(new int[] {i, j});
                    visited[i][j] = true;
                    break;
                }
            }
            if (!q.isEmpty()) break;
        }
        
        int[][] mv = new int[][] {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        while(!q.isEmpty()) {
            int sz = q.size();
            for(int i = 0; i < sz; i++) {
                int[] p = q.poll();
                
                for(int[] n : mv) {
                    int nx = p[0] + n[0];
                    int ny = p[1] + n[1];
                    if (nx < 0 || ny < 0 || nx >= r || ny >= c || visited[nx][ny] || grid[nx][ny] == 'X') continue;
                    if (grid[nx][ny] == '#') return res + 1;
                    q.offer(new int[] {nx, ny});
                    visited[nx][ny] = true;
                }
            }
            res++;
        }
        
        return -1;
    }
}

File: 1730-shortest-path-to-get-food/1730-shortest-path-to-get-food.scala
import scala.collection.mutable._

object Solution {
    def getFood(grid: Array[Array[Char]]): Int = {
        val queue: Queue[(Int, Int)] = Queue()
        
        val r = grid.length
        val c = grid(0).length
        
        for(i <- 0 to r - 1) {
            for (j <- 0 to c - 1) {
                if (grid(i)(j) == '*') queue.enqueue((i, j))
            }
        }
        
        var len = 0
        var found = false
        
        while(!found && queue.nonEmpty) {
            len += 1
            val size = queue.size
            
            for(i <- 1 to size) {
                val p = queue.dequeue
                Seq((1, 0), (-1, 0), (0, 1), (0, -1)).foreach { move =>
                    val nx = p._1 + move._1
                    val ny = p._2 + move._2
                    
                    if (nx >= 0 && ny >= 0 && nx < r && ny < c) {
                        if (grid(nx)(ny) == '#') {
                            found = true
                        } else if (grid(nx)(ny) =='O') {
                            grid(nx)(ny) = 'x'
                            queue.enqueue((nx, ny))
                        }
                    }
                }
            }
        }
        
        if (!found) -1 else len
    }
}